substrate/frame/revive/src/exec.rs

```diff
@@ -120,6 +120,22 @@ impl Key {
 	}
 }
 
+/// Level of reentrancy protection.
+/// This needs to be specifed when a contract makes a message call. This way the calling contract
+/// can specify the level of re-entrancy protection while the callee (and it's recursive callees) is
+/// executing.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum ReentrancyProtection {
+	/// Don't activate reentrancy protection
+	AllowReentry,
+	// Activate strict reentrancy protection. The direct callee and none of its own recursive
+	// callees must be the calling contract.
+	Strict,
+	// Activate reentrancy protection where the direct callee can be the same contract as the
+	// caller but none of the recursive callees of the callee must be the caller
+	AllowNext,
```

Comment:
this is just an extra safety protection for when we grant the stipend right,
maybe we should mention that here

Reply:
I elaborated the comment in https://github.com/paritytech/polkadot-sdk/pull/10166/. It now reads:

```
Activate reentrancy protection where the direct callee can be the same contract as the
caller but none of the recursive callees of the callee must be the caller.

This is used for calls that transfer value but restrict gas so that the callee only has a
stipend gas amount. In Ethereum that is not sufficient for the callee to make another call.
However, due to gas scale differences that guarantee does not automatically hold in revive
and we enforce it explicitly here.
```